File: multiplexer/threaded_server_test.py
"""BaseThreadedMultiplexerServer against a real multiplexer: serial order
with one worker, parallel handling with four, a reply from another thread
later, the search answered while every worker is busy unless told to
decline, a full queue dropping, a handler that raises, draining, and
leaving.
"""

import threading
import time
import unittest

from multiplexer.Multiplexer_pb2 import BackendForPacketSearch
from multiplexer.clients import BackendError, Client
from multiplexer.multiplexer_constants import peers, types
from multiplexer.mxclient import OperationTimedOut
from multiplexer.testing import BackendThread, Cluster, TestClient, wait_until
from multiplexer.threaded_client import ThreadedClient
from multiplexer.threaded_server import BaseThreadedMultiplexerServer, Request

REQUEST = types.PYTHON_TEST_REQUEST
RESPONSE = types.PYTHON_TEST_RESPONSE


class Scripted(BaseThreadedMultiplexerServer):
    """A backend the tests steer by payload: "block" waits for `release`,
    "wait" joins a barrier, "later" is answered by another thread, "raise"
    raises, "event" gets no reply, anything else is upper-cased."""

    multiplexer_client_type = peers.PYTHON_TEST_SERVER

    def __init__(self, addresses, **kwargs):
        super().__init__(addresses, **kwargs)
        self.handled: list[bytes] = []
        self.release = threading.Event()
        self.barrier = threading.Barrier(4, timeout=10)
        self.kept: list[Request] = []
        self.keep_serving_after_error = True

    def handle_message(self, request: Request) -> None:
        payload = request.mxmsg.message
        self.handled.append(payload)
        if payload == b"block":
            self.release.wait(10)
            request.no_response()
        elif payload == b"wait":
            self.barrier.wait()
            request.reply(b"passed", type=RESPONSE)
        elif payload == b"later":
            self.kept.append(request)  # answered by the test, from its thread
        elif payload == b"raise":
            raise ValueError("as asked")
        elif payload == b"close":
            self.close()  # from a worker: an error, not a deadlock
        elif payload.startswith(b"event"):
            request.no_response()
        else:
            request.reply(payload.upper(), type=RESPONSE)

    def on_handler_exception(self, exc: Exception) -> bool:
        return self.keep_serving_after_error


class ThreadedServerTest(unittest.TestCase):
    """One multiplexer for the class, a server per test."""

    @classmethod
    def setUpClass(cls):
        cls.cluster = Cluster(1).__enter__()

    @classmethod
    def tearDownClass(cls):
        cls.cluster.__exit__(None, None, None)

    def serve(self, **kwargs) -> tuple[BackendThread, Scripted]:
        """A Scripted server on its own thread, registered; stopped at the end of the test."""
        served = BackendThread(lambda: Scripted(self.cluster.endpoints, **kwargs)).start()
        self.cluster.wait_for_peer(peers.PYTHON_TEST_SERVER)
        self.addCleanup(lambda: served.stop() if served.running else None)
        assert isinstance(served.backend, Scripted)
        return served, served.backend

    def test_one_worker_handles_in_arrival_order_and_answers(self):
        _, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            lane = client.lane()
            for index in range(100):
                client.send(b"event-%03d" % index, REQUEST, multiplexer=lane)
            self.assertEqual(b"HELLO", client.query(b"hello", REQUEST, multiplexer=lane).message)
        self.assertEqual([b"event-%03d" % index for index in range(100)] + [b"hello"], server.handled)
        wait_until(lambda: server.pending == 0, 10, "the worker done with the query it answered")

    def test_four_workers_handle_four_requests_at_once(self):
        _, server = self.serve(workers=4)
        client = ThreadedClient(self.cluster.endpoints, type=peers.PYTHON_TEST_CLIENT)
        try:
            replies = []
            done = threading.Semaphore(0)
            for _ in range(4):
                client.query(
                    b"wait", REQUEST, timeout=10, callback=lambda result: (replies.append(result), done.release())
                )
            for _ in range(4):
                self.assertTrue(done.acquire(timeout=15), "a request never came back: the barrier was not passed")
            self.assertEqual([b"passed"] * 4, [reply.message for reply in replies])
            self.assertFalse(server.barrier.broken)
        finally:
            client.shutdown()

    def test_a_request_may_be_answered_later_from_another_thread(self):
        _, server = self.serve()

        def answer_later():
            request = wait_until(lambda: server.kept and server.kept[0], 10, "the kept request")
            time.sleep(0.3)
            request.reply(b"finally", type=RESPONSE)

        thread = threading.Thread(target=answer_later)
        thread.start()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            self.assertEqual(b"finally", client.query(b"later", REQUEST, timeout=10).message)
        thread.join()

    def search(self, client: Client, timeout: float):
        """A client's search for a backend of the request type, by hand: the reply or the exception."""
        query = BackendForPacketSearch()
        query.packet_type = REQUEST
        message = client.new_message(message=query, type=types.BACKEND_FOR_PACKET_SEARCH)
        return client.send_and_receive(message, multiplexer=Client.ALL, handle_delivery_errors=True, timeout=timeout)[0]

    def test_the_search_is_answered_while_every_worker_is_busy(self):
        _, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            client.send(b"block", REQUEST)
            client.send(b"event-queued", REQUEST)
            wait_until(lambda: server.pending == 2, 10, "the worker busy and one request waiting")
            self.assertEqual(types.PING, self.search(client.client, 5).type, "answered at once, from the io thread")
            server.release.set()
            self.assertEqual(b"AFTER", client.query(b"after", REQUEST).message)

    def test_declining_the_search_when_full_is_an_option(self):
        _, server = self.serve(decline_searches_when_full=True)
        with TestClient(self.cluster, peers.WEBSITE) as client:
            client.send(b"block", REQUEST)
            wait_until(lambda: server.pending == 1, 10, "the worker busy")
            self.assertEqual(types.PING, self.search(client.client, 5).type, "busy but nothing waiting: answered")
            client.send(b"event-queued", REQUEST)
            wait_until(lambda: server.pending == 2, 10, "one request waiting")
            with self.assertRaises(OperationTimedOut):
                self.search(client.client, 1.0)
            server.release.set()
            self.assertEqual(b"AFTER", client.query(b"after", REQUEST).message)
            self.assertEqual(types.PING, self.search(client.client, 5).type, "idle again: answered")

    def test_a_full_queue_drops_and_the_rest_is_handled_in_order(self):
        _, server = self.serve(queue_size=2)
        with TestClient(self.cluster, peers.WEBSITE) as client:
            lane = client.lane()
            client.send(b"block", REQUEST, multiplexer=lane)
            wait_until(lambda: server.pending == 1, 10, "the worker busy")
            for index in range(5):
                client.send(b"event-%d" % index, REQUEST, multiplexer=lane)
            # A flushing send only says the bytes left; the drops say the rest arrived.
            wait_until(lambda: server.pending == 3 and server.dropped == 3, 10, "two waiting, three dropped")
            server.release.set()
            self.assertEqual(b"MARKER", client.query(b"marker", REQUEST, multiplexer=lane).message)
        self.assertEqual([b"block", b"event-0", b"event-1", b"marker"], server.handled)

    def test_a_raising_handler_reports_backend_error_and_serves_on(self):
        served, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            with self.assertRaises(BackendError) as raised:
                client.query(b"raise", REQUEST)
            self.assertIn(b"as asked", raised.exception.args[0])
            self.assertEqual(b"STILL", client.query(b"still", REQUEST).message)
            server.keep_serving_after_error = False
            with self.assertRaises(BackendError):
                client.query(b"raise", REQUEST)
        with self.assertRaises(ValueError):
            served.stop()  # serve_forever() raised what the handler raised

    def test_draining_declines_searches_and_finishes_the_queue(self):
        served, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            client.send(b"block", REQUEST)
            for index in range(3):
                client.send(b"event-%d" % index, REQUEST)
            wait_until(lambda: server.pending == 4, 10, "the worker busy and three waiting")
            server.start_draining()
            with self.assertRaises(OperationTimedOut):
                self.search(client.client, 1.0)
            server.release.set()
            served.stop()
        self.assertEqual([b"block", b"event-0", b"event-1", b"event-2"], server.handled, "the queue was finished")
        self.cluster.wait_for_peer_gone(peers.PYTHON_TEST_SERVER)

    def test_close_from_a_handler_is_an_error_not_a_deadlock(self):
        served, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            with self.assertRaises(BackendError) as raised:
                client.query(b"close", REQUEST, timeout=10)
            self.assertIn(b"worker thread", raised.exception.args[0])
            self.assertEqual(b"STILL", client.query(b"still", REQUEST).message, "still serving")
        self.assertTrue(served.running)

    def test_stop_from_a_handler_and_the_instance_id(self):
        served, server = self.serve()
        with TestClient(self.cluster, peers.WEBSITE) as client:
            reply = client.query(b"who", REQUEST, to=server.instance_id)
            self.assertEqual((b"WHO", server.instance_id), (reply.message, reply.from_))
        server.stop()
        wait_until(lambda: not served.running, 10, "serve_forever() returned")
        self.cluster.wait_for_peer_gone(peers.PYTHON_TEST_SERVER)


if __name__ == "__main__":
    unittest.main()

File: multiplexer/backend/threaded_server_test.cc
// BaseThreadedMultiplexerServer against a Server in this process: serial
// order with one worker, four requests at once with four, a reply from
// another thread later, the search answered while busy unless told to
// decline, a full queue dropping, a handler that throws, draining.
#include <gtest/gtest.h>

#include <atomic>
#include <condition_variable>
#include <future>
#include <mutex>
#include <thread>

#include "multiplexer/backend/base_threaded_multiplexer_server.h"
#include "multiplexer/client.h"
#include "multiplexer/in_process_multiplexer.h"
#include "multiplexer/multiplexer.constants.h"
#include "multiplexer/threaded_client.h"

using multiplexer::ThreadedClient;
using multiplexer::backend::BaseThreadedMultiplexerServer;
using multiplexer::backend::RequestPtr;
using multiplexer::backend::ThreadedServerOptions;
using multiplexer::testing::InProcessMultiplexer;

namespace {

// A backend the tests steer by payload: "block" waits for release(),
// "wait" joins a barrier of four, "later" is kept for the test to answer,
// "throw" throws, "event..." gets no reply, anything else is upper-cased.
struct Scripted : BaseThreadedMultiplexerServer {
  Scripted(unsigned short port, const ThreadedServerOptions& options = ThreadedServerOptions())
      : BaseThreadedMultiplexerServer({{"127.0.0.1", port}}, multiplexer::peers::PYTHON_TEST_SERVER, options) {}

  void handle_message(const RequestPtr& request) override {
    const std::string payload = request->mxmsg().message();
    {
      std::lock_guard<std::mutex> lock(mutex);
      handled.push_back(payload);
    }
    if (payload == "block") {
      std::unique_lock<std::mutex> lock(mutex);
      released_cv.wait_for(lock, std::chrono::seconds(10), [this] { return released; });
      request->no_response();
    } else if (payload == "wait") {
      std::unique_lock<std::mutex> lock(mutex);
      if (++waiting == 4) {
        released_cv.notify_all();
      } else {
        released_cv.wait_for(lock, std::chrono::seconds(10), [this] { return waiting >= 4; });
      }
      request->reply("passed", multiplexer::types::PYTHON_TEST_RESPONSE);
    } else if (payload == "later") {
      std::lock_guard<std::mutex> lock(mutex);
      kept = request;
    } else if (payload == "throw") {
      throw std::runtime_error("as asked");
    } else if (payload == "close") {
      close();  // from a worker: an error, not a deadlock
    } else if (payload.rfind("event", 0) == 0) {
      request->no_response();
    } else {
      std::string upper = payload;
      for (char& character : upper) {
        character = std::toupper(static_cast<unsigned char>(character));
      }
      request->reply(upper, multiplexer::types::PYTHON_TEST_RESPONSE);
    }
  }
  bool on_handler_exception(const std::exception&) override { return keep_serving; }

  void release() {
    std::lock_guard<std::mutex> lock(mutex);
    released = true;
    released_cv.notify_all();
  }
  std::vector<std::string> snapshot() {
    std::lock_guard<std::mutex> lock(mutex);
    return handled;
  }

  std::mutex mutex;
  std::condition_variable released_cv;
  bool released = false;
  int waiting = 0;
  std::vector<std::string> handled;
  RequestPtr kept;
  std::atomic<bool> keep_serving{true};
};

// A Scripted served on its own thread, as a program would.
struct Served {
  explicit Served(unsigned short port, const ThreadedServerOptions& options = ThreadedServerOptions())
      : server(port, options), thread([this] {
          try {
            server.serve_forever(0.05f);
          } catch (...) {
            failure = std::current_exception();
          }
        }) {}
  ~Served() {
    server.stop();
    thread.join();
  }
  Scripted server;
  std::exception_ptr failure;
  std::thread thread;
};

// A synchronous client: sends events and requests, searches by hand.
struct Requester {
  explicit Requester(unsigned short port) : client(multiplexer::peers::WEBSITE) {
    client.connect("127.0.0.1", port, 5);
  }
  multiplexer::MultiplexerMessage message(const std::string& payload, std::uint32_t type) {
    multiplexer::MultiplexerMessage msg;
    msg.set_id(client.random64());
    msg.set_from(client.instance_id());
    msg.set_type(type);
    msg.set_message(payload);
    return msg;
  }
  void send(const std::string& payload, multiplexer::LanePtr lane = multiplexer::LanePtr()) {
    client.send(message(payload, multiplexer::types::PYTHON_TEST_REQUEST), 5, lane);
  }
  std::string query(const std::string& payload, multiplexer::LanePtr lane = multiplexer::LanePtr()) {
    return client.query(message(payload, multiplexer::types::PYTHON_TEST_REQUEST), 10, lane).third->message();
  }
  // The search clients use to find a backend: PING back, or a timeout.
  std::uint32_t search(float timeout) {
    multiplexer::BackendForPacketSearch search;
    search.set_packet_type(multiplexer::types::PYTHON_TEST_REQUEST);
    multiplexer::MultiplexerMessage msg =
        message(search.SerializeAsString(), multiplexer::types::BACKEND_FOR_PACKET_SEARCH);
    client.flush(client.schedule_one(msg), 5);
    for (;;) {
      multiplexer::IncomingMessage got = client.read_raw_message(timeout);
      if (got.third->references() == msg.id()) {
        return got.third->type();
      }
    }
  }
  multiplexer::Client client;
};

template <typename Predicate>
bool eventually(Predicate predicate, float seconds = 10) {
  std::chrono::steady_clock::time_point deadline =
      std::chrono::steady_clock::now() + std::chrono::milliseconds(static_cast<long>(seconds * 1000));
  while (!predicate()) {
    if (std::chrono::steady_clock::now() > deadline) {
      return false;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
  }
  return true;
}

}  // namespace

TEST(ThreadedServer, OneWorkerHandlesInArrivalOrderAndAnswers) {
  InProcessMultiplexer mx;
  Served served(mx.port);
  Requester requester(mx.port);
  multiplexer::LanePtr lane(new multiplexer::Lane());
  std::vector<std::string> expected;
  for (int index = 0; index < 100; ++index) {
    requester.send("event-" + std::to_string(index), lane);
    expected.push_back("event-" + std::to_string(index));
  }
  EXPECT_EQ("HELLO", requester.query("hello", lane));
  expected.push_back("hello");
  EXPECT_EQ(expected, served.server.snapshot());
  EXPECT_TRUE(eventually([&] { return served.server.pending() == 0; })) << "the worker done with the query";
}

TEST(ThreadedServer, FourWorkersHandleFourRequestsAtOnce) {
  InProcessMultiplexer mx;
  ThreadedServerOptions options;
  options.workers = 4;
  Served served(mx.port, options);
  ThreadedClient client(multiplexer::peers::PYTHON_TEST_CLIENT);
  ASSERT_TRUE(client.connect("127.0.0.1", mx.port, 5));
  std::vector<std::future<ThreadedClient::Result>> results;
  for (int index = 0; index < 4; ++index) {
    results.push_back(std::async(std::launch::async,
                                 [&] { return client.query("wait", multiplexer::types::PYTHON_TEST_REQUEST, 10); }));
  }
  for (auto& result : results) {
    ThreadedClient::Result r = result.get();
    ASSERT_EQ(ThreadedClient::REPLIED, r.outcome) << "the four were not handled at once";
    EXPECT_EQ("passed", r.reply.third->message());
  }
}

TEST(ThreadedServer, ARequestMayBeAnsweredLaterFromAnotherThread) {
  InProcessMultiplexer mx;
  Served served(mx.port);
  std::thread answerer([&] {
    ASSERT_TRUE(eventually([&] {
      std::lock_guard<std::mutex> lock(served.server.mutex);
      return static_cast<bool>(served.server.kept);
    }));
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    served.server.kept->reply("finally", multiplexer::types::PYTHON_TEST_RESPONSE);
  });
  Requester requester(mx.port);
  EXPECT_EQ("finally", requester.query("later"));
  answerer.join();
}

TEST(ThreadedServer, TheSearchIsAnsweredWhileBusyUnlessToldToDecline) {
  InProcessMultiplexer mx;
  {
    Served served(mx.port);
    Requester requester(mx.port);
    requester.send("block");
    requester.send("event-queued");
    ASSERT_TRUE(eventually([&] { return served.server.pending() == 2; }));
    EXPECT_EQ(multiplexer::types::PING, requester.search(5)) << "answered at once, from the io thread";
    served.server.release();
    EXPECT_EQ("AFTER", requester.query("after"));
  }
  ThreadedServerOptions options;
  options.decline_searches_when_full = true;
  Served served(mx.port, options);
  Requester requester(mx.port);
  requester.send("block");
  ASSERT_TRUE(eventually([&] { return served.server.pending() == 1; }));
  EXPECT_EQ(multiplexer::types::PING, requester.search(5)) << "busy but nothing waiting: answered";
  requester.send("event-queued");
  ASSERT_TRUE(eventually([&] { return served.server.pending() == 2; }));
  EXPECT_THROW(requester.search(1), multiplexer::Client::OperationTimedOut) << "saturated: declined";
  served.server.release();
  EXPECT_EQ("AFTER", requester.query("after"));
}

TEST(ThreadedServer, AFullQueueDropsAndTheRestIsHandledInOrder) {
  InProcessMultiplexer mx;
  ThreadedServerOptions options;
  options.queue_size = 2;
  Served served(mx.port, options);
  Requester requester(mx.port);
  multiplexer::LanePtr lane(new multiplexer::Lane());
  requester.send("block", lane);
  ASSERT_TRUE(eventually([&] { return served.server.pending() == 1; }));
  for (int index = 0; index < 5; ++index) {
    requester.send("event-" + std::to_string(index), lane);
  }
  // A flushing send only says the bytes left; the drops say the rest arrived.
  ASSERT_TRUE(eventually([&] { return served.server.pending() == 3 && served.server.dropped() == 3; }));
  served.server.release();
  EXPECT_EQ("MARKER", requester.query("marker", lane));
  EXPECT_EQ((std::vector<std::string>{"block", "event-0", "event-1", "marker"}), served.server.snapshot());
}

TEST(ThreadedServer, AThrowingHandlerReportsBackendErrorAndServesOn) {
  InProcessMultiplexer mx;
  Served served(mx.port);
  Requester requester(mx.port);
  multiplexer::IncomingMessage reply =
      requester.client.query(requester.message("throw", multiplexer::types::PYTHON_TEST_REQUEST), 10);
  EXPECT_EQ(multiplexer::types::BACKEND_ERROR, reply.third->type());
  EXPECT_EQ("as asked", reply.third->message());
  EXPECT_EQ("STILL", requester.query("still"));
  served.server.keep_serving = false;
  reply = requester.client.query(requester.message("throw", multiplexer::types::PYTHON_TEST_REQUEST), 10);
  EXPECT_EQ(multiplexer::types::BACKEND_ERROR, reply.third->type());
  served.thread.join();                // serve_forever() returned on its own, rethrowing
  served.thread = std::thread([] {});  // the destructor joins again
  EXPECT_TRUE(served.failure) << "serve_forever() should have rethrown";
}

TEST(ThreadedServer, CloseFromAHandlerIsAnErrorNotADeadlock) {
  InProcessMultiplexer mx;
  Served served(mx.port);
  Requester requester(mx.port);
  multiplexer::IncomingMessage reply =
      requester.client.query(requester.message("close", multiplexer::types::PYTHON_TEST_REQUEST), 10);
  EXPECT_EQ(multiplexer::types::BACKEND_ERROR, reply.third->type());
  EXPECT_NE(std::string::npos, reply.third->message().find("worker thread"));
  EXPECT_EQ("STILL", requester.query("still")) << "still serving";
}

TEST(ThreadedServer, DrainingDeclinesSearchesAndFinishesTheQueue) {
  InProcessMultiplexer mx;
  Scripted server(mx.port);
  Requester requester(mx.port);
  requester.send("block");
  for (int index = 0; index < 3; ++index) {
    requester.send("event-" + std::to_string(index));
  }
  ASSERT_TRUE(eventually([&] { return server.pending() == 4; }));
  server.start_draining();
  EXPECT_THROW(requester.search(1), multiplexer::Client::OperationTimedOut);
  server.release();
  server.serve_forever(0.05f);  // drained at once: finishes the queue and returns
  EXPECT_EQ((std::vector<std::string>{"block", "event-0", "event-1", "event-2"}), server.snapshot());
}
